[aurora-au] new port

andre-nguyen opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] When updating the upstream version, the `"port-version"` is reset (removed from `vcpkg.json`).
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

andre-nguyen committed:
[au] new port

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

JonLiu1993 commented on ports/au/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+  "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
+  "name": "au",
+  "version-semver": "0.3.5",
+  "description": "A C++14-compatible physical units library with no dependencies and a single-file delivery option. Emphasis on safety, accessibility, performance, and developer experience.",
```
```suggestion
  "description": "A C++14-compatible physical units library with no dependencies and a single-file delivery option. Emphasis on safety, accessibility, performance, and developer experience.",
  "homepage": "https://github.com/aurora-opensource/au",
```

JonLiu1993 commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO aurora-opensource/au
+    REF 0.3.5
```
```suggestion
    REF "${VERSION}"
```

JonLiu1993 commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+vcpkg_cmake_install()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
+file(INSTALL "${SOURCE_PATH}/LICENSE.txt" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE.txt")

```

JonLiu1993 commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+    HEAD_REF main
+    PATCHES cmakelists.patch
+)
+
```
Could you explain why you don't use the `vcpkg_cmake_configure `function?

JonLiu1993 commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+    PATCHES cmakelists.patch
+)
+
+vcpkg_cmake_configure(SOURCE_PATH ${SOURCE_PATH})
```
```suggestion
vcpkg_cmake_configure(SOURCE_PATH “${SOURCE_PATH}”)
```

JonLiu1993 commented on ports/au/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
```
You haven't used the function "vcpkg_cmake_config_fixup", so this is not needed

JonLiu1993 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

andre-nguyen commented:
@JonLiu1993 Thank you for your review. I've gone ahead and opened a PR on the docs site https://github.com/microsoft/vcpkg-docs/pull/405 to include some of the best practices you are showing here.

JonLiu1993 commented:
Tested usage successfully by `au:x64-windows`:
````
au is header-only and can be used from CMake via:

  find_path(AU_INCLUDE_DIRS "au/apply_magnitude.hh")
  target_include_directories(main PRIVATE ${AU_INCLUDE_DIRS})
````

dg0yt commented on ports/au/cmakelists.patch:
```diff
@@ -0,0 +1,62 @@
+-# For Windows: Prevent overriding the parent project's compiler/linker settings
+-set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
+-
+-FetchContent_MakeAvailable(googletest)
+-include(GoogleTest)
```
It is more or less enough to remove those two lines.

andre-nguyen replied:
I see, so that's the line that does the download, not the declare line.

dg0yt commented on ports/au/cmakelists.patch:
```diff
@@ -0,0 +1,62 @@
+   )
+-
+-  # Add the test, if requested.
+-  if (DEFINED ARG_GTEST_SRCS)
```
Changing this line to `if(0)` would be sufficient.

andre-nguyen replied:
Thanks, patch looks much smaller now

dg0yt commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+    REF "${VERSION}"
+    SHA512 4aa3282f6b76fbadd04ca572734f72c86b1b0b4e85fc21a03d1ab00b83d3aea319ab2dac3934361b5f6fa7c4a0dccece94fe0a57f3d73d208315b51b1950e374
+    HEAD_REF main
+    PATCHES cmakelists.patch
```
Nit-pick:
```suggestion
    PATCHES
        cmakelists.patch
```

andre-nguyen replied:
yup

dg0yt commented on ports/au/cmakelists.patch:
```diff
@@ -0,0 +1,62 @@
+ # (This is necessary for it to be usable in other CMake projects.)
+ 
+-set(AU_CMAKE_DIR ${CMAKE_INSTALL_LIBDIR}/cmake/Au)
++set(AU_CMAKE_DIR ${CMAKE_INSTALL_DATAROOTDIR}/cmake/Au)
```
This would be solved without patching with `vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/Au)` (from the `vpckg-cmake-config` host port).

`share/cmake/Au` is also wrong, vcpkg wants `share/au`. (`vcpkg_cmake_config_fixup` would do it right, given the port name.)

andre-nguyen replied:
Thanks, that works, although I then needed to add 
```
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")
```
because `vcpkg_cmake_config_fixup` left behind an empty `lib` folder.

dg0yt commented on ports/au/portfile.cmake:
```diff
@@ -0,0 +1,14 @@
+vcpkg_cmake_configure(SOURCE_PATH "${SOURCE_PATH}")
+vcpkg_cmake_install()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
```
Wait, is this header-only? Then we don't need to build debug.
```suggestion
```
here, and add the top of the file, add
~~~cmake
set(VCPKG_BUILD_TYPE release)  # header-only
~~~

andre-nguyen replied:
Looks like this works. 👍🏼

dg0yt reviewed (changes requested):
The manifest lacks a `license` field.

The patch could be minimized.

Commit:
Add blank line

Commit:
update version

JavierMatosD commented:
Since this port is attempting to take a two letter name, I'm tagging vcpkg-team-review to discuss whether or not there is a need for disambiguation.

andre-nguyen commented:
@JavierMatosD just wondering, what is typically the review turnaround time?
Not urgent, just wondering.

JavierMatosD commented:
@ras0219-msft, @data-queue, @BillyONeal, @AugP and I discussed this today.

@andre-nguyen, we ask that you prepend 'Aurora-' to the port name (`aurora-au`) to disambiguate the port name. 

@chiphogg, + courtesy ping

JavierMatosD commented:
> @JavierMatosD just wondering, what is typically the review turnaround time? Not urgent, just wondering.

We typically hold weekly team meetings to review PRs that need more detailed attention. However, we often receive more PRs than we can reasonably review in a week, so things may occasionally pile up. While there isn’t a “typical” turnaround time, we’re actively working on streamlining the review process to improve the experience for everyone.

chiphogg reviewed (commented):
Wow, this is really exciting!

Independently, other contributors have started the process of bringing Au into conan.  That process is suggesting some CMake changes that I think would help the vcpkg port as well, namely, making it easier to avoid the googletest dependency natively without patching.

I'm excited to see the library reach a wider audience, and grateful for the support!

andre-nguyen committed:
[aurora-au] rename au to aurora-au

andre-nguyen committed:
[aurora-au] update version to 0.3.5

andre-nguyen commented:
Hello @JonLiu1993, just want to check in and see if you are aware the PR has been updated and is ready for another review.

JonLiu1993 commented:
@andre-nguyen, When I tested the usage by `aurora-au:x64-windows`, I get this error, please take a look:
````
1> [CMake] CMake Error at F:/Feature-test/aurora-au/scripts/buildsystems/vcpkg.cmake:859 (_find_package):
1> [CMake]   By not providing "Findgoogletest.cmake" in CMAKE_MODULE_PATH this project
1> [CMake]   has asked CMake to find a package configuration file provided by
1> [CMake]   "googletest", but CMake did not find one.
1> [CMake] 
1> [CMake]   Could not find a package configuration file provided by "googletest"
1> [CMake]   (requested version 1.12.1) with any of the following names:
1> [CMake] 
1> [CMake]     googletestConfig.cmake
1> [CMake]     googletest-config.cmake
1> [CMake] 
1> [CMake]   Add the installation prefix of "googletest" to CMAKE_PREFIX_PATH or set
1> [CMake]   "googletest_DIR" to a directory containing one of the above files.  If
1> [CMake]   "googletest" provides a separate development package or SDK, be sure it has
1> [CMake]   been installed.
1> [CMake] Call Stack (most recent call first):
1> [CMake]   C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.29/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
1> [CMake]   F:/Feature-test/aurora-au/installed/x64-windows/share/aurora-au/AuConfig.cmake:42 (find_dependency)
1> [CMake]   F:/Feature-test/aurora-au/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeProject1/CMakeLists.txt:6 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
````

chiphogg commented:
> @andre-nguyen, When I tested the usage by `aurora-au:x64-windows`, I get this error, please take a look:
> 
> ```
> 1> [CMake] CMake Error at F:/Feature-test/aurora-au/scripts/buildsystems/vcpkg.cmake:859 (_find_package):
> 1> [CMake]   By not providing "Findgoogletest.cmake" in CMAKE_MODULE_PATH this project
> 1> [CMake]   has asked CMake to find a package configuration file provided by
> 1> [CMake]   "googletest", but CMake did not find one.
> 1> [CMake] 
> 1> [CMake]   Could not find a package configuration file provided by "googletest"
> 1> [CMake]   (requested version 1.12.1) with any of the following names:
> 1> [CMake] 
> 1> [CMake]     googletestConfig.cmake
> 1> [CMake]     googletest-config.cmake
> 1> [CMake] 
> 1> [CMake]   Add the installation prefix of "googletest" to CMAKE_PREFIX_PATH or set
> 1> [CMake]   "googletest_DIR" to a directory containing one of the above files.  If
> 1> [CMake]   "googletest" provides a separate development package or SDK, be sure it has
> 1> [CMake]   been installed.
> 1> [CMake] Call Stack (most recent call first):
> 1> [CMake]   C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.29/Modules/CMakeFindDependencyMacro.cmake:76 (find_package)
> 1> [CMake]   F:/Feature-test/aurora-au/installed/x64-windows/share/aurora-au/AuConfig.cmake:42 (find_dependency)
> 1> [CMake]   F:/Feature-test/aurora-au/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
> 1> [CMake]   CMakeProject1/CMakeLists.txt:6 (find_package)
> 1> [CMake] -- Configuring incomplete, errors occurred!
> ```

Possible places to check for this:

- https://github.com/aurora-opensource/au/blob/030298cd9b2d20df7b3cce4256e1151d6afdb677/cmake/AuConfig.cmake.in#L18
- https://github.com/aurora-opensource/au/blob/030298cd9b2d20df7b3cce4256e1151d6afdb677/CMakeLists.txt#L40-L49

dg0yt commented:
There is a patch to disable the googletest dependency on port build. 
This patch must be extended to remove the `find_dependency(googletest)`.

I would also suggest to rename the patch to `disable-googletest.diff` or similar.

andre-nguyen commented:
Thank you everyone for the tips.
Seems to be working on my end now with a small test project.

- I have renamed the patch file to disable-googletest.patch
- Added a usage file

JonLiu1993 commented:
Tested usage successfully by `aurora-au:x64-windows`:
````
The package aurora-au provides CMake targets:

    find_package(Au REQUIRED)
    target_link_libraries(main PRIVATE Au::au)
````

Commit:
Add an empty line

Commit:
update version

chiphogg commented:
Thanks @andre-nguyen for adding this!

When I look at the [official entry on vcpkg](https://vcpkg.io/en/package/aurora-au), it says "Documentation: Unavailable".  We actually have a full doc website, for both the [listed version 0.3.5](https://aurora-opensource.github.io/au/0.3.5/), and for [head](https://aurora-opensource.github.io/au/main/).  How hard would it be to add these links?

andre-nguyen commented:
> Thanks @andre-nguyen for adding this!
> 
> When I look at the [official entry on vcpkg](https://vcpkg.io/en/package/aurora-au), it says "Documentation: Unavailable". We actually have a full doc website, for both the [listed version 0.3.5](https://aurora-opensource.github.io/au/0.3.5/), and for [head](https://aurora-opensource.github.io/au/main/). How hard would it be to add these links?

Looks fairly straightforward, its just a [documentation field in the json file](https://learn.microsoft.com/en-us/vcpkg/reference/vcpkg-json#documentation). I can do it later.

chiphogg commented:
@andre-nguyen, would you be able/interested in updating to Au 0.4.0?  It's out now, and it should avoid the need for a lot of the CMake workarounds!

(I thought I'd ask you first because I've never used vcpkg, never put together a vcpkg PR, and haven't signed the CLA.)

JonLiu1993 commented:
> @andre-nguyen, would you be able/interested in updating to Au 0.4.0? It's out now, and it should avoid the need for a lot of the CMake workarounds!
> 
> (I thought I'd ask you first because I've never used vcpkg, never put together a vcpkg PR, and haven't signed the CLA.)

I commit a PR #42679 update aurora-au to the 0.4.0
